File: app/src/main/java/com/teixeira/subtitles/activities/project/VideoHandlerActivity.java
/*
 * This file is part of SubTypo.
 *
 * SubTypo is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SubTypo is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with SubTypo.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira.subtitles.activities.project;

import android.content.res.Configuration;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import androidx.media3.common.Player;
import com.blankj.utilcode.util.SizeUtils;
import com.blankj.utilcode.util.ThreadUtils;
import com.teixeira.subtitles.fragments.sheets.SubtitleEditorSheetFragment;
import com.teixeira.subtitles.models.Subtitle;
import com.teixeira.subtitles.preferences.Preferences;
import com.teixeira.subtitles.utils.UiUtils;
import com.teixeira.subtitles.R;
import com.teixeira.subtitles.utils.VideoUtils;
import java.util.List;

/**
 * Base class for ProjectActivity that handles most video related things.
 *
 * @author Felipe Teixeira
 */
abstract class VideoHandlerActivity extends BaseProjectActivity {

  private Runnable progressTracker;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    progressTracker = this::progressTracker;

    if (!Preferences.isDevelopmentUndoAndRedoEnabled()) {
      binding.controllerContent.redo.setVisibility(View.GONE);
      binding.controllerContent.undo.setVisibility(View.GONE);
    }
    subtitlesViewModel.setUndoManagerEnabled(Preferences.isDevelopmentUndoAndRedoEnabled());

    binding.videoContent.videoView.setVideoPath(project.getVideoPath());
    binding.videoContent.videoView.seekTo(videoViewModel.getCurrentPosition());
    setVideoViewModelObservers();
  }

  @Override
  protected void onPause() {
    super.onPause();

    videoViewModel.pauseVideo();
  }

  @Override
  protected void preDestroy() {
    super.preDestroy();

    binding.videoContent.videoView.release();
    videoViewModel.setPrepared(false);
  }

  @Override
  protected void configureListeners() {
    super.configureListeners();
    binding.videoContent.videoView.setPlayerListener(
        new Player.Listener() {
          @Override
          public void onPlaybackStateChanged(int state) {
            if (state == Player.STATE_READY) {
              onVideoPrepared(binding.videoContent.videoView.getPlayer());
            } else if (state == Player.STATE_ENDED) {
              binding.controllerContent.play.setImageResource(R.drawable.ic_play);
            }
          }

          @Override
          public void onPositionDiscontinuity(
              Player.PositionInfo oldPosition, Player.PositionInfo newPosition, int reason) {
            videoViewModel.setCurrentPosition(newPosition.contentPositionMs, false);
          }

          @Override
          public void onIsPlayingChanged(boolean isPlaying) {
            if (isPlaying && progressTracker != null) {
              ThreadUtils.getMainHandler().post(progressTracker);
            }
          }
        });

    binding.controllerContent.seekBar.setOnSeekBarChangeListener(
        new SeekBar.OnSeekBarChangeListener() {

          private boolean wasPlaying;

          @Override
          public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            if (fromUser) {
              videoViewModel.setCurrentPosition(progress, true);
            }
          }

          @Override
          public void onStartTrackingTouch(SeekBar seekBar) {
            wasPlaying = videoViewModel.isPlaying();
            if (wasPlaying) videoViewModel.pauseVideo();
          }

          @Override
          public void onStopTrackingTouch(SeekBar seekBar) {
            if (wasPlaying) videoViewModel.playVideo();
          }
        });

    binding.controllerContent.play.setOnClickListener(
        v -> videoViewModel.setPlaying(!videoViewModel.isPlaying()));

    binding.controllerContent.undo.setOnClickListener(v -> subtitlesViewModel.undo());
    binding.controllerContent.redo.setOnClickListener(v -> subtitlesViewModel.redo());
    binding.controllerContent.skipBackward.setOnClickListener(
        v -> binding.videoContent.videoView.seekBackward());
    binding.controllerContent.skipFoward.setOnClickListener(
        v -> binding.videoContent.videoView.seekFoward());
    binding.controllerContent.addSubtitle.setOnClickListener(
        v -> {
          videoViewModel.pauseVideo();
          SubtitleEditorSheetFragment.newInstance(videoViewModel.getCurrentPosition())
              .show(getSupportFragmentManager(), null);
        });
  }

  protected void updateVideoUI(long currentPosition) {
    binding.controllerContent.currentVideoPosition.setText(VideoUtils.getTime(currentPosition));
    binding.controllerContent.seekBar.setProgress((int) currentPosition);
    binding.timeLine.setCurrentPosition(currentPosition);

    List<Subtitle> subtitles = subtitlesViewModel.getSubtitles();
    boolean subtitleFound = false;
    for (int i = 0; i < subtitles.size(); i++) {
      Subtitle subtitle = subtitles.get(i);
      long startTime = VideoUtils.getMilliSeconds(subtitle.getStartTime());
      long endTime = VideoUtils.getMilliSeconds(subtitle.getEndTime());

      if (currentPosition >= startTime && currentPosition <= endTime) {
        binding.videoContent.subtitleView.setSubtitle(subtitle);
        binding.videoContent.subtitleView.setVisibility(View.VISIBLE);
        subtitlesViewModel.setVideoSubtitleIndex(i);
        subtitleFound = true;
        break;
      }
    }

    if (!subtitleFound) {
      binding.videoContent.subtitleView.setVisibility(View.GONE);
      subtitlesViewModel.setVideoSubtitleIndex(-1);
    }
  }

  private void onVideoPrepared(Player player) {
    long duration = player.getDuration();

    binding.controllerContent.videoDuration.setText(VideoUtils.getTime(duration));
    binding.controllerContent.seekBar.setMax((int) duration);
    binding.timeLine.setDuration(duration);
    videoViewModel.setDuration(duration);

    int width = player.getVideoSize().width;
    int height = player.getVideoSize().height;

    if (width > height) {
      binding.videoContent.videoView.getLayoutParams().width = ViewGroup.LayoutParams.MATCH_PARENT;
      binding.videoContent.videoView.requestLayout();
    } else if (height > width
        && getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT) {
      binding.videoContent.getRoot().getLayoutParams().height = SizeUtils.dp2px(350f);
      binding.videoContent.getRoot().requestLayout();
    }
    
    updateVideoUI(videoViewModel.getCurrentPosition());
    videoViewModel.setPrepared(true);
  }

  private void setVideoViewModelObservers() {
    videoViewModel.observeIsPrepared(
        this,
        isPrepared -> {
          UiUtils.setImageEnabled(binding.controllerContent.skipBackward, isPrepared);
          UiUtils.setImageEnabled(binding.controllerContent.play, isPrepared);
          UiUtils.setImageEnabled(binding.controllerContent.skipFoward, isPrepared);
          UiUtils.setImageEnabled(binding.controllerContent.addSubtitle, isPrepared);
        });
    videoViewModel.observeCurrentPosition(
        this,
        currentPositionPair -> {
          long currentVideoPosition = currentPositionPair.first;
          boolean seekTo = currentPositionPair.second;
          if (seekTo) {
            binding.videoContent.videoView.seekTo(currentVideoPosition);
          }
          updateVideoUI(currentVideoPosition);
        });

    videoViewModel.observeIsPlaying(
        this,
        isPlaying -> {
          if (isDestroying()) {
            return;
          }
          binding.controllerContent.play.setImageResource(
              isPlaying ? R.drawable.ic_pause : R.drawable.ic_play);
          binding.videoContent.videoView.setPlaying(isPlaying);
          requireSubtitleListAdapter().setVideoPlaying(isPlaying);
        });
  }

  private void progressTracker() {

    if (isDestroying()) {
      return;
    }

    videoViewModel.setCurrentPosition(binding.videoContent.videoView.getCurrentPosition(), false);

    if (videoViewModel.isPlaying() && progressTracker != null) {
      ThreadUtils.getMainHandler().post(progressTracker);
    }
  }
}

File: app/src/main/java/com/teixeira/subtitles/activities/project/VideoHandlerActivity.kt
/*
 * This file is part of SubTypo.
 *
 * SubTypo is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SubTypo is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with SubTypo.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira.subtitles.activities.project

import android.content.res.Configuration
import android.os.Bundle
import android.view.ViewGroup
import android.widget.SeekBar
import androidx.media3.common.Player
import com.blankj.utilcode.util.SizeUtils
import com.blankj.utilcode.util.ThreadUtils
import com.teixeira.subtitles.R
import com.teixeira.subtitles.subtitle.utils.TimeUtils.getFormattedTime

/**
 * Base class for ProjectActivity that handles most video related things.
 *
 * @author Felipe Teixeira
 */
abstract class VideoHandlerActivity : ProjectHandlerActivity() {

  private var progressTracker: Runnable? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    progressTracker = Runnable { progressTracker() }

    binding.videoContent.videoView.setVideoPath(project.videoPath)
    binding.videoContent.videoView.seekTo(videoViewModel.currentPosition)
    setVideoViewModelObservers()
    configureListeners()
  }

  override fun onPause() {
    super.onPause()

    videoViewModel.pauseVideo()
  }

  override fun preDestroy() {
    super.preDestroy()

    binding.videoContent.videoView.release()
    videoViewModel.isPrepared = false
  }

  private fun configureListeners() {
    binding.videoContent.videoView.setPlayerListener(
      object : Player.Listener {
        override fun onPlaybackStateChanged(state: Int) {
          if (state == Player.STATE_READY) {
            onVideoPrepared(binding.videoContent.videoView.player!!)
          } else if (state == Player.STATE_ENDED) {
            binding.controllerContent.play.setImageResource(R.drawable.ic_play)
            videoViewModel.setCurrentPosition(0, true)
            videoViewModel.pauseVideo()
          }
        }

        override fun onPositionDiscontinuity(
          oldPosition: Player.PositionInfo,
          newPosition: Player.PositionInfo,
          reason: Int,
        ) {
          videoViewModel.setCurrentPosition(newPosition.contentPositionMs, false)
        }

        override fun onIsPlayingChanged(isPlaying: Boolean) {
          if (isPlaying) {
            progressTracker?.also { ThreadUtils.getMainHandler().post(it) }
          }
        }
      }
    )

    binding.controllerContent.seekBar.setOnSeekBarChangeListener(
      object : SeekBar.OnSeekBarChangeListener {

        private var wasPlaying = false

        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
          if (fromUser) videoViewModel.setCurrentPosition(progress.toLong(), true)
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {
          wasPlaying = videoViewModel.isPlaying
          if (wasPlaying) videoViewModel.pauseVideo()
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
          if (wasPlaying) videoViewModel.playVideo()
        }
      }
    )

    binding.controllerContent.play.setOnClickListener {
      videoViewModel.isPlaying = !videoViewModel.isPlaying
    }

    binding.controllerContent.undo.setOnClickListener { subtitlesViewModel.undo() }
    binding.controllerContent.redo.setOnClickListener { subtitlesViewModel.redo() }
    binding.controllerContent.skipBackward.setOnClickListener {
      binding.videoContent.videoView.seekBackward()
    }
    binding.controllerContent.skipFoward.setOnClickListener {
      binding.videoContent.videoView.seekFoward()
    }

    binding.controllerContent.changeLanguage.setOnClickListener { showSubtitleSelectorDialog() }
    binding.controllerContent.addParagraph.setOnClickListener {
      videoViewModel.pauseVideo()
      if (subtitlesViewModel.subtitles.isEmpty()) {
        showSubtitleEditorSheet()
        return@setOnClickListener
      }
      showParagraphEditorSheet()
    }
  }

  protected fun updateVideoUI(currentPosition: Long) {
    requireParagraphListAdapter().setVideoPosition(currentPosition)
    binding.controllerContent.currentVideoPosition.setText(currentPosition.getFormattedTime())
    binding.videoContent.subtitleView.setVideoPosition(currentPosition)
    binding.controllerContent.seekBar.setProgress(currentPosition.toInt())
    binding.timeLine.setCurrentPosition(currentPosition)
  }

  private fun onVideoPrepared(player: Player) {
    val duration = player.duration

    binding.controllerContent.videoDuration.setText(duration.getFormattedTime())
    binding.controllerContent.seekBar.setMax(duration.toInt())
    binding.timeLine.setDuration(duration)
    videoViewModel.duration = duration

    val width = player.videoSize.width
    val height = player.videoSize.height

    when {
      width > height -> {
        binding.videoContent.videoView.getLayoutParams().width = ViewGroup.LayoutParams.MATCH_PARENT
        binding.videoContent.videoView.requestLayout()
      }
      height > width &&
        resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT -> {
        binding.videoContent.root.getLayoutParams().height = SizeUtils.dp2px(350f)
        binding.videoContent.root.requestLayout()
      }
      else -> {}
    }

    updateVideoUI(videoViewModel.currentPosition)
    videoViewModel.isPrepared = true
  }

  private fun setVideoViewModelObservers() {
    videoViewModel.observeIsPrepared(this) { isPrepared ->
      binding.controllerContent.skipBackward.isEnabled = isPrepared
      binding.controllerContent.play.isEnabled = isPrepared
      binding.controllerContent.skipFoward.isEnabled = isPrepared
      binding.controllerContent.addParagraph.isEnabled = isPrepared
    }
    videoViewModel.observeCurrentPosition(this) { (currentPosition, seekTo) ->
      if (seekTo) {
        binding.videoContent.videoView.seekTo(currentPosition)
      }
      updateVideoUI(currentPosition)
    }

    videoViewModel.observeIsPlaying(this) { isPlaying ->
      if (isDestroying) {
        return@observeIsPlaying
      }
      binding.controllerContent.play.setImageResource(
        if (isPlaying) R.drawable.ic_pause else R.drawable.ic_play
      )
      binding.videoContent.videoView.setPlaying(isPlaying)
      requireParagraphListAdapter().isVideoPlaying = isPlaying
    }
  }

  private fun progressTracker() {

    if (isDestroying) {
      return
    }

    videoViewModel.setCurrentPosition(binding.videoContent.videoView.getCurrentPosition(), false)

    if (videoViewModel.isPlaying) {
      progressTracker?.also { ThreadUtils.getMainHandler().post(it) }
    }
  }
}
